java/ql/src/experimental/Security/CWE/CWE-522/InsecureBasicAuth.ql

```diff
@@ -0,0 +1,176 @@
+/**
+ * @name Insecure basic authentication
+ * @description Basic authentication only obfuscates username/password in Base64 encoding, which can be easily recognized and reversed. Transmission of sensitive information not over HTTPS is vulnerable to packet sniffing.
+ * @kind path-problem
+ * @id java/insecure-basic-auth
+ * @tags security
+ *       external/cwe-522
+ *       external/cwe-319
+ */
+
+import java
+import semmle.code.java.frameworks.Networking
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * The Java class `org.apache.http.client.methods.HttpRequestBase`. Popular subclasses include `HttpGet`, `HttpPost`, and `HttpPut`.
+ * And the Java class `org.apache.http.message.BasicHttpRequest`.
+ */
+class ApacheHttpRequest extends RefType {
+  ApacheHttpRequest() {
+    this
+        .getASourceSupertype*()
+        .hasQualifiedName("org.apache.http.client.methods", "HttpRequestBase") or
+    this.getASourceSupertype*().hasQualifiedName("org.apache.http.message", "BasicHttpRequest")
+  }
+}
+
+/**
+ * Class of Java URL constructor.
+ */
+class URLConstructor extends ClassInstanceExpr {
+  URLConstructor() { this.getConstructor().getDeclaringType() instanceof TypeUrl }
+
+  Expr stringArg() {
+    // URLs constructed with any of the four string constructors below:
+    // URL(String spec), URL(String protocol, String host, int port, String file), URL(String protocol, String host, int port, String file, URLStreamHandler handler), URL(String protocol, String host, String file)
+    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+    result = this.getArgument(0) // First argument contains the protocol part.
+  }
+}
+
+/**
+ * The type `java.net.URLConnection`.
+ */
+class TypeHttpUrlConnection extends RefType {
+  TypeHttpUrlConnection() {
+    this.getASourceSupertype*().hasQualifiedName("java.net", "HttpURLConnection")
+  }
+}
+
+/**
+ * String of HTTP URLs not in private domains.
+ */
+class HttpString extends StringLiteral {
+  HttpString() {
+    // Match URLs with the HTTP protocol and without private IP addresses to reduce false positives.
+    exists(string s | this.getRepresentedString() = s |
+      s = "http"
+      or
+      s.matches("http://%") and
+      not s.matches("%/localhost%") and
+      not s.matches("%/127.0.0.1%") and
+      not s.matches("%/10.%") and
+      not s.matches("%/172.16.%") and
+      not s.matches("%/192.168.%")
```

Comment:
Maybe these should check that they are actually matching the host. Currently they would also match (i.e. cause false negatives) for `http://google.com/10.`. Similarly they could erroneously match subdomains, e.g. `http://10.google.com`
So what about using the following regexes?
- [Matching `localhost`](https://regex101.com/r/n9WDgr/2)
- [Matching IPv4 addresses](https://regex101.com/r/n1QQCw/1)
- [Matching IPv6 addresses](https://regex101.com/r/MWytkn/1)

Reply:
I've updated the query to check the protocol+host pattern with both IPv4 and IPv6 addresses.